chef-config/lib/chef-config/path_helper.rb

```diff
@@ -201,7 +201,11 @@ def self.escape_glob_dir(*parts)
     end
 
     def self.relative_path_from(from, to, windows: ChefUtils.windows?)
-      Pathname.new(cleanpath(to, windows: windows)).relative_path_from(Pathname.new(cleanpath(from, windows: windows)))
+      if windows
+        Pathname.new(cleanpath(to, windows: windows)).relative_path_from(Pathname.new(cleanpath(from, windows: windows)))
+      else
+        Pathname.new(cleanpath(Pathname.new(to).relative_path_from(Pathname.new(from))))
+      end
```

Comment:
Several questions:
1. if you can call cleanpath on the result, why not do that in both cases, rather than in sub-paths in the windows case?
2. YOu mentioned in your last comment why didn't use cleanpath, but you did, and so I'm confused.
3. What is "hack" vs "new" in your benchmarks?

Reply:
> 3. What is "hack" vs "new" in your benchmarks?

I've renamed the benchmark methods to (what I hope are) more understandable names, v2 being the current implementation in this PR, v1 being the original speedup.

```
ruby 3.1.4p223 (2023-03-30 revision 957bb7cb81) [x86_64-linux]
Warming up --------------------------------------
            orig_rpf     1.192k i/100ms
              v2_rpf   637.000 i/100ms
              v1_rpf     2.075k i/100ms
Calculating -------------------------------------
            orig_rpf     12.392k (±20.3%) i/s   (80.70 μs/i) -    118.008k in  10.021897s
              v2_rpf     15.833k (±22.6%) i/s   (63.16 μs/i) -    147.784k in  10.007126s
              v1_rpf     22.029k (±12.9%) i/s   (45.40 μs/i) -    215.800k in  10.000513s

Comparison:
              v1_rpf:    22028.5 i/s
              v2_rpf:    15833.1 i/s - same-ish: difference falls within error
            orig_rpf:    12391.8 i/s - 1.78x  slower
```

> 2. YOu mentioned in your last comment why didn't use cleanpath, but you did, and so I'm confused.

I had [copied the gsub from cleanpath](https://gist.github.com/dafyddcrosby/13d5fa4db708bc7e0386a426f9c232cf#file-benchmark_relative_path_from-rb-L16-L18) to avoided the unnecessary [Pathname.new at the start of the method definition](https://github.com/chef/chef/blob/e224c63341d754142516084cc4cff1aeef0ca800/chef-config/lib/chef-config/path_helper.rb#L152). When it came up in the first PR review session where we looked at v1, I was asked to make the gsub in cleanpath in-place (which I did with #14579) and continue using the cleanpath command in relative_path_from instead of copying in cleanpath, so that's what I did for v2.

1. if you can call cleanpath on the result, why not do that in both cases, rather than in sub-paths in the windows case?

The sub-paths are because Windows can have strange edge-cases with DOS-ish paths. I'm not a Windows expert, so I left that alone. `relative_path_from` is supposed to return a Pathname object, whereas [cleanpath returns a String](https://github.com/chef/chef/blob/e224c63341d754142516084cc4cff1aeef0ca800/chef-config/lib/chef-config/path_helper.rb#L152). Annotating a bit shows why, I think.

```ruby
if windows
  Pathname.new(cleanpath(to, windows: windows)). # Pathname object with string argument
    relative_path_from( # Pathname method
      Pathname.new( # Pathname
      cleanpath(from, windows: windows) # string argument
      ))
else
  Pathname.new( # Pathname
    cleanpath(  # string argument
      Pathname.new(to). # Pathname
      relative_path_from( # Pathname method
        Pathname.new(from) # Pathname
        )))
end
```

Reply:
@jaymzh Have you had a chance to review this?

Reply:
I'm not sure what you mean by "same-ish: difference falls within error"

12.3k - 15.8k is a 3.5k speed up (orig -> v2)
15.8k - 22k  is another 6k speeds up (v2 -> v1)

Are you saying the 3.5k is within margin of error? If so, the argument is it's not worth doing, unless you do it the v1 way? Or are you saying something else?

What I originally suggested was taking your initial fixes that you applied here when you dropped cleanpath and _apply them to_ cleanpath, which should help all callers, not just this one caller. That would probably yield much more benefit.